File: 1038. Binary Search Tree to Greater Sum Tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func bstToGst(root *TreeNode) *TreeNode {
    helper(root, 0)
    return root
}

func helper(root *TreeNode, value int) int {
    if root == nil { return 0 }
    rightValue := helper(root.Right, value)
    leftValue := helper(root.Left, rightValue + root.Val + value)
    subTreeSum := rightValue + leftValue + root.Val
    root.Val += rightValue + value
    return subTreeSum
}

File: java/1038. Binary Search Tree to Greater Sum Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode bstToGst(TreeNode root) {
        helper(root, 0);
        return root;
    }
    private int helper(TreeNode root, int value) {
        if (root == null) return 0;
        int rightValue = helper(root.right, value);
        int leftValue = helper(root.left, value + root.val + rightValue);
        int subtreeSum = rightValue + leftValue + root.val;
        root.val += value + rightValue;
        return subtreeSum;
    }
}
